utilities/git/hooks/pre-commit.sh

```diff
@@ -14,7 +14,28 @@ then
    exit 1
 fi
 
-clang-format \
-    -i \
-    -style=file \
-    $(git diff --cached --name-only *.cpp *.h)
+# Some of the git tools can set this variable, which can cause problems
+# https://magit.vc/manual/magit/My-Git-hooks-work-on-the-command_002dline-but-not-inside-Magit.html
+unset GIT_LITERAL_PATHSPECS
+
+#Change the delimiter used by 'for' so we can handle spaces in names
+IFS=$'\n'
+for FILE in $(git diff --cached --name-only "*.cpp" "*.h")
+do
+    # Store the current version in case it has been modified
+    TMPFILE="$(mktemp)"
+    cp "$FILE" "$TMPFILE"
+    # Temporarily remove modified versions, if they exist
+    git checkout -q "$FILE"
+    # Format
+    clang-format -i -style=file "$FILE"
+    # Add the formatted file
+    git add "$FILE"
+    # Restore the old, possibly modified version
+    mv "$TMPFILE" "$FILE"
+    # Format the uncached version to keep it in sync with the cached
+    clang-format -i -style=file "$FILE"
+done
+
+# Don't commit if there are no changes to commit
+test "$(git diff --cached)" != ""
```

Comment:
I've approved this and there's no need to change, but since travis is being uncommunicative anyway this can be simplified a bit further:
```suggestion
git diff --cached --exit-code
```

Reply:
I've tried testing this version, but it seems that it doesn't work. 
From the `git-diff` manpages:
```
--exit-code
           Make the program exit with codes similar to diff(1). That is, it exits with 1 if there 
were differences and 0 means no differences.
```
Wouldn't that be the inverse of what we need?

Reply:
I'll merge as is then.

Reply:
Should we consider something like:
```
# Don't commit if there are no changes to commit
if [ "$(git diff --cached --quiet)" != "0" ]; then
    exit 0
else
    exit 1
fi
```